source/common/http/conn_manager_utility.cc

```diff
@@ -95,7 +95,7 @@ Network::Address::InstanceConstSharedPtr ConnectionManagerUtility::mutateRequest
   // peer. Cases where we don't "use remote address" include trusted double proxy where we expect
   // our peer to have already properly set XFF, etc.
   Network::Address::InstanceConstSharedPtr final_remote_address;
-  bool single_xff_address;
+  bool single_xff_address = false;
```

Comment:
hm, I think we need to pass this one through the API too.  Maybe with better naming in the API though, like "allow_trusted_address_checks"?  I think it should be possible to use this plug-in and have the plug-in decide if the original endpoint is trusted WDYT?
(and sorry for not catching this on the first pass)

Reply:
How about something like this:

```
struct OriginalIPDetectionParams {
  // Note: headers will be sanitized after the OriginalIPDetection::detect() call.
  Http::RequestHeaderMap& request_headers;
  // The downstream directly connected address.
  Network::Address::InstanceConstSharedPtr downstream_remote_address; 
};

struct OriginalIPDetectionResult {
  // An address that represents the detected address or nullptr.
  Network::Address::InstanceConstSharedPtr detected_remote_address; 
  // Is the detected address trusted (e.g.: used to determine if it's an internal request).
  bool allow_trusted_address_checks;
};

/**
 * Interface class for original IP detection extensions.
 */
class OriginalIPDetection {
public:
  virtual ~OriginalIPDetection() = default;

  /**
   * Detect the final remote address and it should be trusted.
   *
   * @param param supplies the OriginalIPDetectionParams params for detection.
   */
  virtual OriginalIPDetectionResult detect(struct OriginalIPDetectionParams& params) PURE;
};

```

Reply:
I think that the downstream_remote_address argument is really important as both a way to do trusted direct connected host checks and as a potential fallback if headers that provide an IP are missing.  Do add it to OriginalIPDetectionParams.

Is less clear to me how 'allow_trusted_address_checks' would be used or how it would be provided to filters that need to know if this is "trusted client" or "trusted direct connected host".

One thing that we may want to add: a way to indicate that IP detection failed and the request should be rejected.

Reply:
I _think_ Alyssa means we'd use `allow_trusted_address_checks` (formerly `single_xff_address`) here:

https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_utility.cc#L154

```
  const bool internal_request =
      single_xff_address && final_remote_address != nullptr &&
      config.internalAddressConfig().isInternalAddress(*final_remote_address);
```

no?

Reply:
Ok, yes -- so we could make it such that if `ConnectionManagerUtility::mutateRequestHeaders()` returns nullptr in `ConnectionManagerImpl::ActiveStream::decodeHeaders()`, then the request should be rejected (in a configurable way).  

So we probably want the extension's result to be something like:

```
struct OriginalIPDetectionResult {
  // An address that represents the detected address or nullptr.
  Network::Address::InstanceConstSharedPtr detected_remote_address; 
  // Is the detected address trusted (e.g.: used to determine if it's an internal request).
  bool allow_trusted_address_checks;
  // Should the request be rejected (e.g.: because detection failed).
  bool reject_request;
  // More params about how to reject the request perhaps (status code, reason, etc). 
};
```

This would allow extension implementors to not need an additional companion filter (a good thing).